src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -221,27 +326,153 @@ static ucs_status_t uct_cuda_copy_mem_alloc(uct_md_h md, size_t *length_p,
-                                                     *length_p));
+        status = UCS_ERR_UNSUPPORTED;
+        ucs_carray_for_each(method, methods, num_methods) {
+            switch (*method) {
+            case UCT_CUDA_ALLOC_METHOD_FABRIC:
+                if (!md->config.fabric_supported) {
+                    continue;
+                }
+
+                status = uct_cuda_copy_mem_alloc_fabric(md, alloc_handle);
+                if (status == UCS_OK) {
+                    goto allocated;
+                }
+                break;
+            case UCT_CUDA_ALLOC_METHOD_DEVICE:
+                status = UCT_CUDADRV_FUNC_LOG_ERR(
+                        cuMemAlloc(&alloc_handle->ptr, alloc_handle->length));
+                if (status == UCS_OK) {
+                    goto allocated;
+                }
+                break;
+            default:
+                status = UCS_ERR_UNSUPPORTED;
+                ucs_error("supported allocation methods: fabric, device");
+                goto err;
+            }
+        }
+    } else if (mem_type == UCS_MEMORY_TYPE_CUDA_MANAGED) {
+        status = UCT_CUDADRV_FUNC_LOG_ERR(
+                cuMemAllocManaged(&alloc_handle->ptr, alloc_handle->length,
+                                  CU_MEM_ATTACH_GLOBAL));
     } else {
-        status =
-            UCT_CUDADRV_FUNC_LOG_ERR(cuMemAllocManaged((CUdeviceptr*)address_p,
-                                                       *length_p,
-                                                       CU_MEM_ATTACH_GLOBAL));
+        ucs_error("allocation mem_types supported: cuda, cuda-managed");
+        return UCS_ERR_INVALID_PARAM;
     }
 
+err:
     if (status != UCS_OK) {
+        ucs_free(alloc_handle);
         return status;
     }
 
-    *memh_p = *address_p;
+allocated:
+    *memh_p    = alloc_handle;
+    *address_p = (void*)alloc_handle->ptr;
+    *length_p  = alloc_handle->length;
     return UCS_OK;
 }
 
+static ucs_status_t
+uct_cuda_copy_mem_release_fabric(uct_cuda_copy_alloc_handle_t *alloc_handle)
+{
+#if HAVE_CUDA_FABRIC
+    ucs_status_t status;
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(
+            cuMemRelease(alloc_handle->generic_handle));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(
+            cuMemUnmap(alloc_handle->ptr, alloc_handle->length));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    return UCT_CUDADRV_FUNC_LOG_ERR(
+            cuMemAddressFree(alloc_handle->ptr, alloc_handle->length));
+#else
+    return UCS_ERR_UNSUPPORTED;
+#endif
+}
+
+static unsigned
+uct_cuda_copy_detect_vmm(void *address, ucs_memory_type_t *vmm_mem_type)
+{
+    unsigned is_vmm            = 0;
+#if HAVE_CUDA_FABRIC
+    ucs_status_t status        = UCS_OK;
+    CUmemAllocationProp prop   = {};
+    CUmemGenericAllocationHandle alloc_handle;
+#endif
+
+    *vmm_mem_type = UCS_MEMORY_TYPE_UNKNOWN;
```

Comment:
IMO we don't need to set *vmm_mem_type" if the memory is not VMM. so we can move this after checking that cuMemRetainAllocationHandle was successful